### PR TITLE
[Mechanic] Disabled Self-Repair

### PR DESCRIPTION
**Pull request by Azure3141**
**Feature**

This PR addresses the feature described in issues #7266 and #4678 (sorta kinda) and requested by @Ebreus on Discord.

## Summary
This adds a new attribute, `disabled recovery timer` (name TBR) that allows disabled ships to repair themselves after a counter reaches the value of `disabled recovery timer`. It also adds attributes to apply a cost of energy or fuel (which must be met in order for a ship to repair itself) and apply any of the various damage types in a single burst upon a ship self repairing.

## Usage examples
My testing ship was a Hawk, which I shot up until it was disabled (took a couple tries to not instantly kill it).

```
ship "Hawk"
	sprite "ship/hawk"
	thumbnail "thumbnail/hawk"
	attributes
		category "Interceptor"
		"cost" 670000
		"shields" 2500
		"hull" 500
		"required crew" 1
		"bunks" 4
		"mass" 110
		"drag" 2.1
		"heat dissipation" .8
		"fuel capacity" 300
		"cargo space" 20
		"outfit space" 200
		"weapon capacity" 50
		"engine capacity" 70
		"disabled recovery timer" 180
		"disabled recovery heat" 500
		"disabled recovery energy" 100
		"disabled recovery slowing" 100
```

In this case, the Hawk had to have at least 100 energy in order to repair itself, which was subtracted upon the repair actually happening, as well as the addition of 500 heat and 100 slowing.

## Testing Done
Did the aforementioned Hawk abuse. Probably still need to test energy/fuel requirements actually being able to restrict the repair from happening.

## Save File
N/A

## Artwork Checklist
N/A

## Performance Impact
Likely none.

### CODE REVIEW

**Comment by Ebreus** [comment]
Thanks for the awesome mechanic!

**Comment by Zitchas** [comment]
When did the timer start? When disabled? Or when last hit? Personally I would prefer from last hit with a possibility of overriding it to from when disabled. I can think of story situations where I would want the later, but most of the time I think from last hit post being disabled is better.

I do believe there is a conversation on GitHub somewhere for this, too.

File: data/_ui/tooltips.txt [hunk 1]
```diff
@@ -293,6 +293,42 @@ tip "repair delay:"
...
+tip "disabled recovery disruption:"
+	`The dosription damage applied when a disabled ship repairs itself.`
+
+tip "disabled recovery slowing:"
+	`The slowing damage applied when a disabled ship repairs itself.`
+
+tip "disabled recovery discharge:"
+	`The discharge damage applied when a disabled ship repairs itself.`
+
+tip "disabled recovery corrosion:"
```
Comment (Hurleveur) [inline data/_ui/tooltips.txt:324]: does this make sense? wouldnt you end up staying disabled most of the time lol?

**Comment by xX-Dillinger-Xx** [comment]
I left a rather lengthy post over in #7266. I wanted to add my thoughts but I didn't want to flood this PR.

File: data/_ui/tooltips.txt [hunk 2]
```diff
@@ -293,6 +293,42 @@ tip "repair delay:"
...
+	`The heat applied when a disabled ship repairs itself.`
+
+tip "disabled recovery ionization:"
+	`The ion damage applied when a disabled ship repairs itself.`
+
+tip "disabled recovery scrambling:"
+	`The scrambling damage applied when a disabled ship repairs itself.`
+
+tip "disabled recovery disruption:"
+	`The dosription damage applied when a disabled ship repairs itself.`
```
Comment (EjoThims) [inline data/_ui/tooltips.txt:316]:
```suggestion
	`The disruption damage applied when a disabled ship repairs itself.`
```

**Review by EjoThims — CHANGES_REQUESTED** [review]
Just a minor typo.

**Review by EjoThims — APPROVED** [review]

**Comment by Zitchas** [comment]
I am ambivalent about using "last hit" for the player. I think it is better, but not strongly. I do think it is much better for NPCs, though, because we don't have to worry about their feelings, sand having repair work reset by the player or other enemies hitting it makes more sense. Only the biggest and toughest of ships should be successfully recovering from being disabled while they are actively being shot.

Actually, even better would be to have a 'disable recovery reset chance' attribute that survivors what the chances are. Old reliable so like the Firebird? Low chance of disable recovery being reset by a hit. Some fancy delicate rescuing Flivvr? Really high chance of reset.

But honestly, the more things can be put into the player's hands and not just a dictated way the game works, the better. Meaning of they don't bother investing in tools to improve their recovery, then they recovery slowly and increase the risk of a stray hit keeping them disabled for longer. Even better if ways to reduce the chance each hit had of resetting the timer. 

Having A timer is good, but only marginal improvement as most will probably still reload. The more we can make it something that players can actively plan for and make decisions about, the more it becomes just another part of the game to be experienced as part of the player's story.

**Review by OcelotWalrus — APPROVED** [review]

File: data/_ui/tooltips.txt [hunk 3]
```diff
@@ -293,6 +293,42 @@ tip "repair delay:"
 
 tip "disabled repair delay:"
 	`The number of seconds it takes for hull repairs to begin after being disabled.`
+	
+tip "disabled recovery timer:"
```
Comment (TomGoodIdea) [inline data/_ui/tooltips.txt:297]: You have "disabled recovery **time**" in other places, so either correct it here or change it in the code. Although I'd suggest renaming to "disabled recovery delay" to match repair and shield gen delays.

**Comment by quyykk** [comment]
Should this be tied to hull repair in some way (or use a new hull repair attribute) instead of a timer? Having a small hull ship self-repair just as fast as a large hull ship seems weird.

**Comment by xX-Dillinger-Xx** [comment]
> Should this be tied to hull repair in some way (or use a new hull repair attribute) instead of a timer? Having a small hull ship self-repair just as fast as a large hull ship seems weird.

I would think a smaller hull ship should self repair faster. Less to repair, but less crew so maybe crew can be taken into consideration as well . So, I feel it should be a timer tied to the current "hull" amount of each ship.  The lower the hull number and the higher the number of crew, the faster you recover to a "not disabled" state. At that point Self Repair ends and regular shield and hull recoveries kick in, if equipped.  Whole idea was to prevent a player from being stranded, not fully restore their ships.

**Comment by xX-Dillinger-Xx** [comment]
It can't feel like magic. It needs to have some logical explanation as to what's happening.  Figure out and describe how the ship is being repaired, from our characters perspective, then figure out how to represent that in the game. Do that, and it becomes believable. The current 'docking to repair' is enough magic already. It takes several seconds to minutes to disable a ship put it can be fixed in about 1 second. Not believable at all.

**Comment by Hurleveur** [comment]
I dont think it matters, instant repair is just good gameplay and thats always the focus. We can always find things to justify the gameplay we want (in repair case its just standard for all ships bigger than a fighter to have a repair mechanism)

Back to the matter at hand, I think big ships should have to equip more repair outfits to have it be as fast as small ships. Mb just base it on mass or smth

**Comment by Zitchas** [comment]
I would prefer to have the base repair timer specified as a ship data point, with the capacity to have outfits that modify that by percentage, fixed amount, or override. 

While I generally agree that bigger sis should take longer, I consider this to be in the same category of data as "big ships have more shields." Which is generally true, but subject to so many other factors that it can only be decided by creators, not by automated game mechanic.

**Comment by xX-Dillinger-Xx** [comment]
So my last comment was so crazy it doesn't even get a response, just a thumbs down. It's ok if you don't agree with me, I'm starting to expect that around here. It's no secret that there are two distinct groups within the ES community, and we all know what happens to ideas that get suggested by the "non clique" members. I guess it's become clear which camp I belong in. 

I know it's pointless but, I still maintain it can't feel like magic. If it is made instant it's basically the same as making a ship unable to be disabled in the first place. Being **_invincible_** is bad game play. Here is the current suggested game play scenario.
- you fight and get disabled
- **_poof!_** your not disabled 

If you get disabled there should consequences in the form of having to wait.  If you want instant, you as just use the already included "magic go back in time button"  `ESC`

Humans, as well as others, do not have hull repair but some of you are talking about adding 'magic band-aid' repair kits or outfits. 
Think about it, if we don't have hull repair, who's fixing the damage, crew members, and **_that takes time_**. If we do have hull repair, clearly the nanobots where unable to keep up and where depleted and they **_take time_** to replenish.

**Comment by quyykk** [comment]
> So my last comment was so crazy it doesn't even get a response

@Hurleveur did address it: Gameplay over Realism. We have this conversation a lot of times, and people fall on different sides, but usually people can agree that we cannot make everything hyperrealistic (I'm not sure if that's what you were implying, so I apologize if not). I also apologize if it feels that ideas from the "out-group" always get shot down, this is absolutely not the case.

> If you get disabled there should consequences in the form of having to wait.

This suggested change already includes a timer, unless you're talking about the instant repair when you ask a ship for help. In that case, the time it takes for the ship to dock you is already kinda a timer already. I'm not sure if adding another timer on top of that is a good idea from a gameplay perspective.

**Comment by tibetiroka** [comment]
> So my last comment was so crazy it doesn't even get a response, just a thumbs down.

It seems like something is wrong with github on your end, because there are two comments responding to your previous comment? Sometimes the auto-reload feature of github is not reliable; reloading the page tends to help in these cases.

**Comment by xX-Dillinger-Xx** [comment]
@quyykk 
I didn't suggest hyper realism. I said make sure it has a timer that is "explainable" by something more realistic. Every time I mention 'real' everybody blows a gasket.  Sorry, but real is more believable and **_IF_** something can be explained in a believable way it is more real. They go hand in hand so there really isn't sides.  I understand that not everything can be explained but there should always be an attempt to try. It also makes for better game play. 

I know it includes a timer and that was what I was supporting while suggesting that it needs to be tied to something that makes sense. That was your question after all. I simply mentioned that repair from other ships already feels like magic. I never suggested changing it. It is what it is. But, I certainly wouldn't be opposed to repair taking longer then refueling.
As for `"out-group" always get shot down, this is absolutely not the case.` I'm not going to get into that here. 

@Azure3141 
I never said the idea was bad. 
> It'll only be instant if you set the repair delay attribute to 0, which is just bad content design.

This is why I said, it needs to be tied to something. Not just a simple timer. Different ships do need different timers and **should** be handled in this PR. Why add only half a solution just to go back and re-due it again latter? Like I said tie it to the hull and crew or outfit space stats. Just anything would be a contributing factor as to how long it takes to repair to a non disabled state. That's assuming you want it to make sense instead feeling like magic. Your choice, except the challenge or keep it simple.

**Comment by xX-Dillinger-Xx** [comment]
👍

**Comment by Hecter94** [comment]
> I deleted my content as it was not appreciated. Do not take it as done in a fit of rage. I'm sorry if my form of criticism is considered aggressive. I forget some people can't handle criticism.

We appreciate all kinds of feedback, including constructive criticism. 

However, please remember to maintain a professional and respectful tone while sharing your thoughts. GitHub is a hub for professionals actively involved in development, and we expect everyone using the platform to contribute to creating a constructive and productive environment.

As long as you follow the guidelines, please continue to provide valuable feedback. Our ultimate goal is to make Endless Sky the best product possible, and player feedback is crucial to achieving that objective.

**Comment by petervdmeer** [comment]
I've read through #4678, through #7266 and through the thread here; this PR seems to be doing exactly what was requested. (There were a lot of "out of scope" and "unrelated" discussions, but they can go to their own tickets.)

This PR also seems safe for merging; the only time when it is active is when the condition `if(isDisabled && attributes.Get("disabled recovery time"))` is true, and the attributes are only visible when used. The PR is also quite small.

**Pull request merged**